Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py

```diff
@@ -157,16 +188,48 @@ def _process_reservation(self, reservation, threebot_id, source=None, refund=Non
                                        refund=refund)
 
 
-def _validate_reservation(reservation):
-    for field in ['size', 'email']:
+def _validate_reservation_base(reservation):
+    for field in ['email']:
+        if not getattr(reservation, field):
+            raise ValueError("field '%s' cannot be empty" % field)
+
+    if not _validate_location(reservation.location):
+        raise ValueError("location '%s' is not valid" % reservation.location)
+
+
+def _validate_reservation_s3(reservation):
+    for field in ['size']:
         if not getattr(reservation, field):
             raise ValueError("field '%s' cannot be empty" % field)
 
     if reservation.size < 0 or reservation.size > 2:
         raise ValueError('reservation size can only be 1 or 2')
 
-    if not _validate_location(reservation.location):
-        raise ValueError("location '%s' is not valid" % reservation.location)
+
+def _validate_reservation_vm(reservation):
+    for field in ['size']:
+        if not getattr(reservation, field):
+            raise ValueError("field '%s' cannot be empty" % field)
+
+    if reservation.size < 0 or reservation.size > 2:
+        raise ValueError('reservation size can only be 1 or 2')
+
+
+def _validate_reservation_namespace(reservation):
+    for field in ['size']:
+        if not getattr(reservation, field):
+            raise ValueError("field '%s' cannot be empty" % field)
+
+    if not reservation.mode or reservation.mode not in ['seq', 'user', 'direct']:
```

Comment:
Do we need to validate it manually like this? Don't schemas allow to validate this automatically? Seems a bit silly otherwise (not to say error prone).